src/stream/src/executor/watermark_filter.rs

```diff
@@ -0,0 +1,477 @@
+        assert_eq!(
+            watermark_type,
+            input.schema().data_types()[event_time_col_idx]
+        );
+        let mut input = input.execute();
+
+        let first_barrier = expect_first_barrier(&mut input).await?;
+        table.init_epoch(first_barrier.epoch);
+        // The first barrier message should be propagated.
+        yield Message::Barrier(first_barrier);
+
+        // Initiate and yield the first watermark.
+        let mut current_watermark =
+            Self::get_global_max_watermark(&table, watermark_type.clone()).await?;
+
+        yield Message::Watermark(Watermark::new(
+            event_time_col_idx,
+            current_watermark.clone(),
+        ));
+
+        #[for_await]
+        for msg in input {
+            let msg = msg?;
+            match msg {
+                Message::Chunk(chunk) => {
+                    let chunk = chunk.compact();
+
+                    // Empty chunk should not be processed.
+                    if chunk.cardinality() == 0 {
+                        continue;
+                    }
+
+                    let watermark_array = watermark_expr
+                        .eval_infallible(chunk.data_chunk(), |err| {
+                            ctx.on_compute_error(err, &info.identity)
+                        });
+
+                    // Build the expression to calculate watermark filter.
+                    let watermark_filter_expr = Self::build_watermark_filter_expr(
+                        watermark_type.clone(),
+                        event_time_col_idx,
+                        current_watermark.clone(),
+                    )?;
+
+                    // Assign a new watermark.
+                    for watermark in watermark_array.iter() {
+                        // Ignore the Null watermark
+                        if watermark.is_some() && to_datum_ref(&current_watermark) < watermark {
+                            current_watermark = watermark.to_owned_datum();
+                        }
+                    }
+
+                    let pred_output = watermark_filter_expr
+                        .eval_infallible(chunk.data_chunk(), |err| {
+                            ctx.on_compute_error(err, &info.identity)
+                        });
+
+                    if let Some(output_chunk) = SimpleFilterExecutor::filter(chunk, pred_output)? {
+                        yield Message::Chunk(output_chunk);
+                    };
+
+                    yield Message::Watermark(Watermark::new(
+                        event_time_col_idx,
+                        current_watermark.clone(),
+                    ));
+                }
+                Message::Watermark(watermark) => {
+                    if watermark.col_idx != event_time_col_idx {
+                        yield Message::Watermark(watermark)
+                    } else {
+                        bail!("WatermarkFilterExecutor should not receive a watermark on the event it is filtering.")
```

Comment:
Is this true? 🥵 cc @st1page

Reply:
I am not sure. I think in our first version we can ban it in frontend :hot_face: But what is the difficulty to implement it as a complete executor? @yuhao-su

Reply:
I'll print a warning here and handle the watermark.